service/controller/resource/alerting/alertmanagerconfigsecret/resource.go

```diff
@@ -0,0 +1,177 @@
+package alertmanagerconfigsecret
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/giantswarm/k8sclient/v5/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/prometheus-meta-operator/pkg/template"
+	"github.com/giantswarm/prometheus-meta-operator/service/controller/resource/generic"
+	"github.com/giantswarm/prometheus-meta-operator/service/key"
+)
+
+const (
+	Name                     = "alertmanagerconfigsecret"
+	templateDirectory        = "/opt/prometheus-meta-operator"
+	templatePath             = "files/templates/alertmanager/alertmanager.yaml"
+	notificationTemplatePath = "files/templates/alertmanager/notification-templates.tmpl"
+)
+
+type Config struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+
+	Installation     string
+	Provider         string
+	HTTPProxy        string
+	HTTPSProxy       string
+	NoProxy          string
+	OpsgenieKey      string
+	GrafanaAddress   string
+	SlackApiURL      string
+	SlackProjectName string
+	Pipeline         string
+
+	TemplatePath string
+}
+
+type TemplateData struct {
+	Provider         string
+	Installation     string
+	ProxyURL         *string
+	OpsgenieKey      string
+	GrafanaAddress   string
+	SlackApiURL      string
+	SlackProjectName string
+	Pipeline         string
+}
+
+func New(config Config) (*generic.Resource, error) {
+	clientFunc := func(namespace string) generic.Interface {
+		c := config.K8sClient.K8sClient().CoreV1().Secrets(namespace)
+		return wrappedClient{client: c}
+	}
+
+	if config.TemplatePath == "" {
+		config.TemplatePath = path.Join(templateDirectory, templatePath)
+	}
+
+	c := generic.Config{
+		ClientFunc: clientFunc,
+		Logger:     config.Logger,
+		Name:       Name,
+		GetObjectMeta: func(v interface{}) (metav1.ObjectMeta, error) {
+			return getObjectMeta(v, config)
+		},
+		GetDesiredObject: func(v interface{}) (metav1.Object, error) {
+			return toSecret(v, config)
+		},
+		HasChangedFunc: hasChanged,
+	}
+
+	r, err := generic.New(c)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return r, nil
+}
+
+func getObjectMeta(v interface{}, config Config) (metav1.ObjectMeta, error) {
+	cluster, err := key.ToCluster(v)
+	if err != nil {
+		return metav1.ObjectMeta{}, microerror.Mask(err)
+	}
+
+	return metav1.ObjectMeta{
+		Name:      fmt.Sprintf("alertmanager-%s", config.Installation),
+		Namespace: key.Namespace(cluster),
+	}, nil
+}
+
+func toSecret(v interface{}, config Config) (*corev1.Secret, error) {
+	objectMeta, err := getObjectMeta(v, config)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	notificationTemplate, err := ioutil.ReadFile(path.Join(templateDirectory, notificationTemplatePath))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	alertmanagerConfigSecret, err := toData(v, config)
```

Comment:
A nitpick, but it confused me a bit as `toData` returns the contents of the config file to be included in the secret, not the secret itself.

```suggestion
	alertmanagerConfig, err := toData(v, config)
```